ea-2021-04-ndvi-automation.py

```diff
@@ -0,0 +1,683 @@
+    # TODO add tests to ensure the arrays are the same .shape
+    band = rxr.open_rasterio(band_path, masked=True).squeeze()
+
+    # Only run this step if a valid range tuple is provided
+    if valid_range:
+        mask = ((band < valid_range[0]) | (band > valid_range[1]))
+        band = band.where(~xr.where(mask, True, False))
+
+    return band
+
+
+
+def mask_crop_ndvi(all_bands,
+                   crop_bound,
+                   pixel_qa,
+                   vals):
+    """Open and mask a single landsat band using a pixel_qa layer.
+
+    Parameters
+    -----------
+    all_bands : list
+        a list containing the xarray objects for landsat  bands 4 and  5
+    crop_bound: geopandas GeoDataFrame
+        A geopandas dataframe to be used to crop the raster data using rasterio mask().
+    pixel_qa: xarray DataArray
+        An xarray DataArray with pixel qa values that have not yet been turned into a mask (0s and 1s)
+    vals: list
+        A list of values needed to create the cloud mask
+
+
+    Returns
+    -----------
+    ndvi_crop : Xarray Dataset
+        a cropped and masked xarray object containing NDVI values
+    """
+
+    crop_json = crop_bound.geometry
+
+    # Clip pixel qa cloud mask layer
+    cl_mask_crop = pixel_qa.rio.clip(crop_json)
+
+    # Calculate NDVI
+    ndvi_xr = (all_bands[1]-all_bands[0]) / (all_bands[1]+all_bands[0])
+    # Clip NDVI layer
+    ndvi_crop = ndvi_xr.rio.clip(crop_json)
+    
+    # Apply cloud mask to NDVI
+    ndvi_crop = ndvi_crop.where(~pixel_qa.isin(vals))
+
+    return ndvi_crop
+
+
+# In[6]:
+
+
+all_bands_harv = []
+for aband in band_paths:
+    print("Opening up", aband)
+    cleaned_band = open_clean_bands(band_path=aband, valid_range=(0, 10000))
+    all_bands_harv.append(cleaned_band)
+    date = aband[50:58]
+    
+# Open up the cloud mask layer
+
+# Cloud no data vals for Landsat 8 -
+vals = [328, 392, 840, 904, 1350, 352, 368, 416,
+        432, 480, 864, 880, 928, 944, 992, 480, 992]
+
+# Open cloud mask layer
+qa_xr_path = glob(os.path.join(harv_crop_path, "*qa*"))
+
+
+# Clip the cloud mask layer
+cl_mask = rxr.open_rasterio(qa_xr_path[0], masked=True).squeeze()
+
+
+# Get a list of each directory
+path = os.path.join("ndvi-automation", "sites")
+
+
+# Get a list of both site directories
+sites = glob(path + "/*/")
+
+    
+# Open up the shapefile for clipping your landsat data to the study area
+vector_dir_harv = os.path.join(sites[1],
+                          "vector")
+
+site_name = os.path.basename(os.path.normpath(sites[1]))
+    
+# Open crop boundary
+site_boundary_path = os.path.join(vector_dir_harv,  site_name + "-crop.shp")
+crop_bound = gpd.read_file(site_boundary_path)
+    
+ndvi_clean = mask_crop_ndvi(all_bands=all_bands_harv,
+                            crop_bound=crop_bound,
+                            pixel_qa=cl_mask,
+                            vals=vals)
+
+
+mean_ndvi = np.mean(ndvi_clean)
+
+
+harv_date = pd.to_datetime(date, format='%Y%m%d')
```

Comment:
lines 320-360 code looks good, finding the notes a little hard to follow though (that may be just me!)